src/lib/utils/createSendEventForHits.ts

```diff
@@ -0,0 +1,157 @@
+import { InstantSearch, Hit } from '../../types';
+import { InsightsEvent } from '../../middleware/insights';
+
+type BuiltInSendEventForHits = (
+  eventType: string,
+  hits: Hit | Hit[],
+  eventName?: string
+) => void;
+type CustomSendEventForHits = (customPayload: any) => void;
+export type SendEventForHits = BuiltInSendEventForHits & CustomSendEventForHits;
+
+type BuiltInBindEventForHits = (
+  eventType: string,
+  hits: Hit | Hit[],
+  eventName?: string
+) => string;
+type CustomBindEventForHits = (customPayload: any) => string;
+export type BindEventForHits = BuiltInBindEventForHits & CustomBindEventForHits;
+
+type BuildPayload = (options: {
+  widgetType: string;
+  helper: AlgoliaSearchHelper;
+  methodName: 'sendEvent' | 'bindEvent';
+  args: any[];
+}) => InsightsEvent | null;
+
+const buildPayload: BuildPayload = ({
+  helper,
+  widgetType,
+  methodName,
+  args,
+}) => {
+  if (args.length === 1) {
+    return args[0];
+  }
+  const eventType: string = args[0];
+  const hits: Hit | Hit[] = args[1];
+  const eventName: string | undefined = args[2];
+  if (!hits) {
+    if (__DEV__) {
+      throw new Error(
+        `You need to pass hit or hits as the second argument like:
+  ${methodName}(eventType, hit);
+  `
+      );
+    } else {
+      return null;
+    }
+  }
+  if ((eventType === 'click' || eventType === 'conversion') && !eventName) {
+    if (__DEV__) {
+      throw new Error(
+        `You need to pass eventName as the third argument for 'click' or 'conversion' events like:
+  ${methodName}('click', hit, 'Product Purchased');
+
+  To learn more about event naming: https://www.algolia.com/doc/guides/getting-insights-and-analytics/search-analytics/click-through-and-conversions/in-depth/clicks-conversions-best-practices/
+  `
+      );
+    } else {
+      return null;
+    }
+  }
+  const hitsArray = Array.isArray(hits) ? hits : [hits];
+  const queryID = hitsArray[0].__queryID;
+  const objectIDs = hitsArray.map(hit => hit.objectID);
+  const positions = hitsArray.map(hit => hit.__position);
+  const index = helper.getIndex();
+
+  if (eventType === 'view') {
+    return {
+      insightsMethod: 'viewedObjectIDs',
+      widgetType,
+      eventType,
+      payload: {
+        eventName: eventName || 'Hits Viewed',
+        index,
+        objectIDs,
+      },
+    };
+  } else if (eventType === 'click') {
+    return {
+      insightsMethod: 'clickedObjectIDsAfterSearch',
+      widgetType,
+      eventType,
+      payload: {
+        eventName,
+        index,
+        queryID,
+        objectIDs,
+        positions,
+      },
+    };
+  } else if (eventType === 'conversion') {
+    return {
+      insightsMethod: 'convertedObjectIDsAfterSearch',
+      widgetType,
+      eventType,
+      payload: {
+        eventName,
+        index,
+        queryID,
+        objectIDs,
+      },
+    };
+  } else if (__DEV__) {
+    throw new Error(`eventType("${eventType}") is not supported.
```

Comment:
Having different paths execution in dev and prod might lead to unexpected behaviors. Should we throw in all environments or display a warning instead?

Reply:
It's about analytics, so even if something's wrong, it shouldn't throw errors in production just because of analytics. However, I think it's good to throw in __DEV__, so that they can get notified and have a chance to fix things before deployment. 

In the line 107, it's not 

> It could go wrong, please be aware of xxx.

but, it's actually

> It won't send any events at all.

WDYT?

Reply:
Yep it makes sense on a user POV. It just harder to test the two code paths and the flow is harder to follow.